llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp

```diff
@@ -996,10 +996,32 @@ bool Vectorizer::isSafeToMove(
   LLVM_DEBUG(dbgs() << "LSV: isSafeToMove(" << *ChainElem << " -> "
                     << *ChainBegin << ")\n");
 
-  assert(isa<LoadInst>(ChainElem) == IsLoadChain);
+  assert(isa<LoadInst>(ChainElem) == IsLoadChain &&
+         isa<LoadInst>(ChainBegin) == IsLoadChain);
+
   if (ChainElem == ChainBegin)
     return true;
 
+  if constexpr (IsLoadChain) {
```

Comment:
Would the previous SCEV check prevent reaching here in some subset of situations?

Reply:
Yes. However, it's hard to reason about, and I'm starting to think that there may be false positives with that SCEV patch.